test-network-function/networking/suite.go

```diff
@@ -324,3 +348,109 @@ func testNodePort(env *config.TestEnvironment) {
 		}
 	})
 }
+
+func parseVariables(res string, declaredPorts map[key]string) error {
+	var p Port
+	err := json.Unmarshal([]byte(res), &p)
+	if err != nil {
+		return err
+	}
+	for element := range p {
+		var k key
+		k.port = p[element].ContainerPort
+		k.protocol = p[element].Protocol
+		declaredPorts[k] = p[element].Name
+	}
+	return nil
+}
+func declaredPortList(container int, podName, podNamespace string, declaredPorts map[key]string) error {
+	ocCommandToExecute := fmt.Sprintf(commandportdeclared, podName, podNamespace, container)
+	res, _ := utils.ExecuteCommand(ocCommandToExecute, ocCommandTimeOut, interactive.GetContext(false))
+	err := parseVariables(res, declaredPorts)
+	if err != nil {
+		log.Error(err)
+	}
+	return err
+}
+
+func listeningPortList(commandlisten []string, nodeOc *interactive.Context, listeningPorts map[key]string) error {
+	var k key
+	listeningPortCommand := strings.Join(commandlisten, " ")
+	res, err := utils.ExecuteCommand(listeningPortCommand, ocCommandTimeOut, nodeOc)
+	if err != nil {
+		return err
+	}
+	lines := strings.Split(res, "\n")
+	for _, line := range lines {
+		fields := strings.Fields(line)
+		if indexprotocolname > len(fields) || indexport > len(fields) {
+			return err
+		}
+		s := strings.Split(fields[indexport], ":")
+		p, _ := strconv.Atoi(s[1])
+		k.port = p
+		k.protocol = strings.ToUpper(fields[indexprotocolname])
+		listeningPorts[k] = ""
+	}
+	return nil
+}
+
+func checkIfListenIsDeclared(listeningPorts, declaredPorts map[key]string) (flag bool, res map[key]string) {
+	res = make(map[key]string)
+	if len(listeningPorts) == 0 || len(declaredPorts) == 0 {
+		return false, res
+	}
+	flag = true
+	for k := range listeningPorts {
+		_, ok := declaredPorts[k]
+		if !ok {
+			tnf.ClaimFilePrintf(fmt.Sprintf("The port %d on protocol %s in pod %s is not declared.", k.port, k.protocol, listeningPorts[k]))
+			res[k] = listeningPorts[k]
+			flag = false
+		}
+	}
+	return flag, res
+}
+
+func testListenAndDeclared(env *config.TestEnvironment) {
+	declaredPorts := make(map[key]string)
+	listeningPorts := make(map[key]string)
+	undeclaredPorts, res := make(map[key]string), false
+	testID := identifiers.XformToGinkgoItIdentifier(identifiers.TestServicesDoNotUseNodeportsIdentifier)
+	ginkgo.It(testID, func() {
+		for _, podUnderTest := range env.PodsUnderTest {
+			for i := 0; i < podUnderTest.ContainerCount; i++ {
+				err := declaredPortList(i, podUnderTest.Name, podUnderTest.Namespace, declaredPorts)
+				if err != nil {
+					log.Errorf("Failed: There are no declared ports for pod name %s in pod namespace %s", podUnderTest.Name, podUnderTest.Namespace)
+					continue
```

Comment:
when we fail to get declared ports from one of the containers, there is no point to continue with this pod.

- Add this pod to a skipped list so we can check at the end before calling ginkgo.Fail or gomega.Assert(len(skippedPods), 0). 
- This "continue" statement needs to be on the outer for loop (line 421). See [this reference](https://relistan.com/continue-statement-with-labels-in-go)

Reply:
you're right, I fixed it.